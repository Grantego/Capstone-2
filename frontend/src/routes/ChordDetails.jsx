import React from "react";
import { useLoaderData } from "react-router";
import SpellerApi from "../../api";
import "./ChordDetails.css"

export async function loader({params}) {
  let res = await SpellerApi.getChord(encodeURIComponent(params.name))
  return res
}

const ChordDetails = () => {
  const {chord} = useLoaderData()
    return (
        <>
        <h1 id="chord-detail-name">{chord.name}</h1>
        <h2 id="chord-detail-spelling">Spelling: {chord.spelling}</h2>
        <div id="chord-chart-box">
        <img id="chord-detail-img" src={chord.chordChart} alt={`${chord.name} Chord Chart`} />
        </div>

        </>
    )
}

export default ChordDetails