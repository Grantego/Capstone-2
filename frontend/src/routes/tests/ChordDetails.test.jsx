import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import ChordDetails, {loader} from '../ChordDetails';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect, vi } from 'vitest';
import SpellerApi from '../../../api';

vi.mock('../../../api',  () => {
    return {
        default: vi.fn().mockImplementation(() => {
            return {
                getChord: vi.fn().mockResolvedValue({chord: {name: "C", spelling: "C, E, G", chordChart: "testchart"}})                
            }
        })         
    }
});

describe('ChordDetails component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();
    
    const mockLoader = async () => {
        const apiInstance = new SpellerApi();
        const res = await apiInstance.getChord();
        console.log(res)
        return {chord: res.chord};
    }

    router = createMemoryRouter([
        {
            path: '/',
            element: <ChordDetails/>,
            loader: mockLoader
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders all elements correctly", async () => {

    localStorage.setItem('token', 'sampletoken')

    render(<RouterProvider router={router} />);

    await waitFor(() => {
        expect(screen.getByAltText("C Chord Chart")).toBeInTheDocument()
        expect(screen.getByText("Spelling: C, E, G")).toBeInTheDocument()
        expect(screen.getByText("C")).toBeInTheDocument()
        
    })

  })
});