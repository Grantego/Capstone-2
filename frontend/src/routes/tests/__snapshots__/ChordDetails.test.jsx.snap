// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChordDetails component > matches snapshot 1`] = `<DocumentFragment />`;
